### ISSUE TITLE
Double-sent requests

### ISSUE THREAD

**Issue by staltz** [cyclejs/core#238]
_From @staltz on July 17, 2015 9:2_

```js
function main(responses) {
  console.log('This only gets called once.')
  const url = '/ontologies/acl.ttl';
  var request$ = Cycle.Rx.Observable.just(url);
  
  var vtree$ = responses.http
    .do(r => console.log('this gets called twice', r))
    .mergeAll()
    .do(r => console.log('this gets called once', r))
    .map(res => h('p', res.text.substr(0, 100) + '...'));

  return {
    dom: vtree$,
    http: request$
  };
}

let [requests, responses] = Cycle.run(main, {
  dom: makeDOMDriver('#app-container'),
  http: makeHTTPDriver(),
  db: makeLocalDbDriver(),
});
```

_Copied from original issue: cyclejs/cycle-http-driver#9_

**Comment by staltz** [comment]
_From @tchap on January 17, 2016 11:27_

For me it looks like the request is sent 5 times :-)

**Comment by staltz** [comment]
_From @tchap on January 17, 2016 11:27_

I mean, I am sending a single GET request in my code. I am not running this particular example...

**Comment by staltz** [comment]
_From @tchap on January 17, 2016 11:56_

Actually not sure mine is the same case. I am using `mergeAll` and I am seeing everything once on the client, but the server is receiving the same request 5 times...

**Comment by staltz** [comment]
_From @tchap on January 17, 2016 12:2_

Yeah, seeing 5 responses before `mergeAll`...

**Comment by staltz** [comment]
_From @tchap on January 17, 2016 14:9_

Was trying to find the issue. Not sure why, but while `createResponse$` is called exactly once, the callback in `Rx.Observable.create` in the same function is called 5 times.

**Comment by staltz** [comment]
_From @tchap on January 17, 2016 14:17_

I mean, the create function itself is called 5 times. Totally weird...

**Comment by staltz** [comment]
_From @tchap on January 17, 2016 19:55_

Patched this lib to wrap `superagent` in a promise and then used `Rx.Observable.fromPromise` and it started working properly...

**Comment by staltz** [comment]
_From @arnodenuijl on January 22, 2016 20:39_

I think I had the same problem. One request going in to the driver and two requests being made to the server. laszlokorte explained on gitter that the function ```function createResponse$(reqOptions)``` in the driver creates a cold observable. So every subscription to the ```response$``` results in an extra call being made. Work around is to call ```share()``` on the ```response$``` you get from the driver to make the observable hot **or** to make the request ```eager``` which has the 'side' effect to make the ```response$``` hot.

If you had five requests going out then it looks like you had five effective subscriptions. I think wrapping the call in a promise has the same work around effect (making the response$ hot). 

I would expect the response$ to always be hot. Can't think of a situation where this 'cold' behavior is useful. (but maybe someone else can?!).


See http://arnodenuijl.github.io/CycleDoubleHttp/index.html for a working example where I had two calls being made off of one request. Just open the network tab in the developer tools and click the button. You will see two requests being made for every click. The source code is on https://github.com/arnodenuijl/CycleDoubleHttp/blob/master/src/app.ts. In my case I used the HTTP ```response$``` for generating the ```state$``` **(line 35)**. And the ```state$``` was used for generating the ```DOM``` **(line 42)** **and** for creating the HTTP ```request$``` **(line 69)**. 

By using the ```state$``` in two places, one going into the DOM driver and one going into the HTTP driver,  two subscriptions are effective and two requests are being made.

**Comment by staltz** [comment]
_From @tchap on January 23, 2016 14:17_

@arnodenuijl Cheers, will take a look when I have time.

**Comment by staltz** [comment]
_From @laszlokorte on January 23, 2016 14:36_

@arnodenuijl Now that I think about it the reason for the observable being cold is pretty clear: If the request fails, a cold observable can retry the request just by subscribing again. With a hot observable you can not retry the request (without creating a whole new one which might not be possible if you do not know all the parameters or do not have access to the request sink or if the one finally consuming the response does not know about cycle but just consumes failable observables)

**Comment by staltz** [comment]
> I would expect the response$ to always be hot. Can't think of a situation where this 'cold' behavior is useful. (but maybe someone else can?!).

@arnodenuijl Like everything else with RxJS, cold is the default. I actually agree with you that I expect hot as default, but this is just how RxJS is, always cold by default. And there are deep implications for that. Cold is more general than Hot, so there's a reason why it's default. It's useful specially in cases where you need to retry, like Laszlo said.

So rather, than going against the RxJS normal behavior, I would stick to keeping the cold behavior, and instructing people to `.share()` those. Cycle may soon get support for multiple stream libraries, and then those may allow a hot-by-default approach.

**Comment by vladap** [comment]
I did some small POC with Cycle some time ago. I remember I had some problem using .share on responses from HTTPDriver and had to use .shareReplay(1) instead. If you would kill me I can't remember how the problem has manifested, so I apologize it is really low value input. Because I'm still Rx noob my naive reasoning at the time what something along the lines that HTTP driver is using Subject somewhere in the way hence it has to use .shareReplay(1) instead of .share().

Anyway it is more likely that the fault was on my side because was using decorator for HTTP Driver to provide some default behaviour for HTTP errors, mainly unauthorized and I was sharing by default. Is something wrong with my code which could cause the above problems?

I'm just pretty sure there was something wrong with using .share() in the code below. I will try to replicate if you won't see a problem. Just I'm finishing a sprint at the moment and still behind so little time.

```
export function createHttpMiddleware(httpDriver, history) {
  return req$ => {
    const handleUnauthorized = err => {
      if (err.response && err.response.statusCode === 401) {
        history.push({pathname: '/login'})
      }
      return Observable.just(err)
    }

    return httpDriver(req$)
      .map(x => {
        let res$ = x.catch(handleUnauthorized)
        if (res$.source.request) {
          res$.request = res$.source.request
        }
        return res$
      })
      .shareReplay(1)
  }
}
```

**Comment by vladap** [comment]
Note, that the pattern is trying to achieve... when you are already logged on some page and your auth expires and you e.g. click a button which generates protected http request you are redirected to login form. It is missing returning back original routed when successfully logged in again.

**Comment by vladap** [comment]
I went through my commit messages in that POC and this one problem was fixed by replacing .share() by .shareReplay(1):

"Table component didn't triggered HTTP request, not receiving data resulting in an empty page."

**Comment by felipecsl** [comment]
It still makes double requests for me, tried both `share()` and `shareReplay(1)`

**Comment by laszlokorte** [comment]
@felipecsl could you show your code?

**Comment by felipecsl** [comment]
@laszlokorte it's on #262

**Comment by ivan-kleshnin** [comment]
> I did some small POC with Cycle some time ago. I remember I had some problem using .share on responses from HTTPDriver and had to use .shareReplay(1) instead. If you would kill me I can't remember how the problem has manifested, so I apologize it is really low value input. 

@vladap `shareReplay(1)` is "required" for streams with `scan()` otherwise they don't broadcast a "current" state on subscribe. Usually it leads to initial state being "missing". So my rule of thumb is to add `share()` to every stream you expose and to add `shareReplay(1)` to every stateful stream. 

The other thing with state is that such streams are closer to Property concept.
So you'll want to `.distinctUntilChanged()` them.

I'm actually playing with a store abstraction:

```js
let store = curry((seed, update) => {
  return update
    .startWith(seed)
    .scan(scanFn)
    .shareReplay(1)
    .distinctUntilChanged();
});
```

**Comment by Cmdv** [comment]
@staltz we still classing this as a bug? implement things to be hot in the API or update docs to show the use of `share() / shareReplay(1)`

**Comment by felipecsl** [comment]
for me, using `eager: true` fixed this, as stated on #262

**Comment by Cmdv** [comment]
@felipecsl ah ok cool, well maybe updating docs + extra tests will solve this issue. I'll add more info to the docs like that there is no confusion and create a test for good measures

**Comment by staltz** [comment]
lets close this issue, but the plan is to make eager be the default. Actually all sources should be multicast observables (hot).

**Comment by Cmdv** [comment]
ok cool, well if we need that in the api create new issue and I'll pick it up :smile:

**Comment by laszlokorte** [comment]
@ivan-kleshnin That's a really good summary and explanation of the usage of scan, share and shareReplay! I would love to see that published somewhere in the cycle or rx docs.

**Comment by ivan-kleshnin** [comment]
@laszlokorte thanx!

I will publish an article on my site and Rx guys may link it somehow if they find it's useful.

**Comment by vladap** [comment]
This share() here and .shareReplay(1) there is quite confusing. I probably understand now one case when to use what but still I don't understand why and what is the technical difference between them. Which leads me to doubts if there are other edge cases. It looks a bit leaky, like I have to know how streams works underneath.

I basically understand this hot/cold thing as caching/memoizing some part of computation. In Apache Spark I call .cache() on its DAG to avoid recomputations, Apache Flink can even detect in its optimizer when it needs to should use it (at least some cases, I haven't study its details yet).

Has been share() and shareReplay(1) somehow reworked in RxJs 5 for better good or is it quite same?

**Comment by ivan-kleshnin** [comment]
I apologize for spamming this thread :disappointed_relieved:  Here is one final post.

> Has been share() and shareReplay(1) somehow reworked in RxJs 5 for better good or is it quite same?

In short: it's getting better.
For example, they renamed `replay` to `publishReplay` which is a good clarification.

I'll try to explain. Everything is quite simple once you get it :smile: 

---

1. RxJS has cold `Observable` and hot `Subject` primitives. Cold is useful for representing finite streams (otherwise they would finish immediately). Hot is better for all(?) other cases... So I would prefer having Hot by default but that's not how RxJS is built. 

2. To make an `Observable` behave (like) hot RxJS *fake it* using `Subject` under the cover. 
    This is the simplest solution which "works"    

3. RxJS exposes a `multicast()` operator which does what desribed in 2.
     But it expects an exact Subject type to work. And there are more than one...

4. So RxJS just predefines two flavors of `multicast`. Just for convenience.
     
   ```js
   publish()       // multicast() with RegularSubject()
   publishReplay() // multicast() with ReplaySubject()
   ```

    If you look at the RxJS sources – those above are almost oneliners (extra code is for polymorphism and other stuff irrelevant for explanation).

5. There is one more step. Hot Observables need to be runned manually (by definition).
    So you run `connect()` on them to start event flow. 

    But often you want to avoid this manual book-keeping and fire a flow
    automatically once you have a first subscriber.

    So RxJs provides `refCount()` operators which does that for you.

   These are also oneliners. One for each code branch above.
   
  ```js
   share()       // publish() with refCount()
   shareReplay() // publishReplay() with refCount()
   ```

To recap

```js
// Cold -> Hot
.multicast() // requires connect()

.publish() // multicast() + regular Subject(), requires connect()
.share()   // publish() + refCount()

.publishReplay() // multicast() + ReplaySubject(), requires connect()
.shareReplay()   // publishReplay() + refCount()
```

---

All this is a bit oversimplified but I hope you've got the point.

**Comment by TylorS** [comment]
@ivan-kleshnin I'm probably just being nit-picky but there is also `publishBehavior()` which is equal to `multicast() + BehaviorSubject()`. Also if you're trying to explain specifically RxJS 5 (?); `shareReplay()` is no longer an operator as far as I can tell.

Thanks for the nice write up, and look forward to your article.

**Comment by ivan-kleshnin** [comment]
>  I'm probably just being nit-picky but there is also publishBehavior() which is equal to multicast() + BehaviorSubject()

Yeah. I just never used that one.

> Also if you're trying to explain specifically RxJS 5 (?); shareReplay() is no longer an operator as far as I can tell. Thanks for the nice write up, and look forward to your article.

No I wasn't trying to be RxJS 5 specific. Thank you too. I'll try to do my best.
